socorro/lib/util.py

```diff
@@ -2,225 +2,229 @@
-          else:
-            self.secondaryLimitedSizeCache.append(y)
-            self.secondaryCacheSize += 1
-        else:
-          self.cache.append(y)
-        yield y
-    #finally:
-    #  self.stopUsingSecondaryCache()
+    def __init__(self, anIterator):
+        self.theIterator = anIterator
+        self.cache = []
+        self.secondaryLimitedSizeCache = collections.deque()
+        self.secondaryCacheMaximumSize = 11
+        self.secondaryCacheSize = 0
+        self.useSecondary = False
+
+    def close(self):
+        self.theIterator.close()
+
+    def __iter__(self):
+        for x in self.theIterator:
+            if self.useSecondary:
+                if self.secondaryCacheSize == self.secondaryCacheMaximumSize:
+                    self.secondaryLimitedSizeCache.popleft()
+                    self.secondaryLimitedSizeCache.append(x)
+                else:
+                    self.secondaryLimitedSizeCache.append(x)
+                    self.secondaryCacheSize += 1
+            else:
+                self.cache.append(x)
+            yield x
+
+    def useSecondaryCache(self):
+        self.useSecondary = True
+
+    def stopUsingSecondaryCache(self):
+        self.useSecondary = False
+        self.cache.extend(self.secondaryLimitedSizeCache)
+        self.secondaryCacheSize = 0
+        self.secondaryLimitedSizeCache = collections.deque()
 
-#-----------------------------------------------------------------------------------------------------------------
-import signal
-# Don't know why this isn't available by importing signal, but not.
-signalNameFromNumberMap = dict( ( (getattr(signal,x),x) for x in dir(signal) if (x.startswith('SIG') and not x.startswith('SIG_')) ) )
 
+class StrCachingIterator(CachingIterator):
+    def __init__(self, anIterator):
+        super(StrCachingIterator, self).__init__(anIterator)
+
+    def __iter__(self):
+        for x in self.theIterator:
+            # warning expecting a '\n' on the end of every line
+            y = repr(x)[1:-3]
+            if self.useSecondary:
+                if self.secondaryCacheSize == self.secondaryCacheMaximumSize:
+                    self.secondaryLimitedSizeCache.popleft()
+                    self.secondaryLimitedSizeCache.append(y)
+                else:
+                    self.secondaryLimitedSizeCache.append(y)
+                    self.secondaryCacheSize += 1
+            else:
+                self.cache.append(y)
+            yield y
+
+
+def flatten(d, parent_key='', sep='.'):
+    """Return a dictionary with only top-level keys. The sub-level fields
+    are moved to a top-level key named by concatenating the keys of each level.
+
+    For example:
+        IN:
+        {
+            "a": 1,
+            "b": {
+                "c": 2,
+                "d": {
+                    "e": 3
+                }
+            }
+        }
+        OUT:
+        {
+            "a": 1,
+            "b.c": 2,
+            "b.d.e": 3
+        }
```

Comment:
I tend to do this either using code blocks or doctests. One advantage of doctests is that you can test them and thus know when your examples are wrong. Both code blocks and doctests will get syntax highlighting when rendered in Sphinx.

I have no idea if this is rendered in Sphinx or not. If it is, it's worth changing.

Something like this:

```
For example:

>>> flatten({'a': 1, 'b': {'c': 2, 'd': {'e': 3}}})
{'a': 1, 'b.c': 2, 'b.d.e': 3}

```